libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -404,7 +404,15 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
 			SecondarySourceLocation ssl;
 			for (auto const* decl: declarations)
 				ssl.append("The shadowed declaration is here:", decl->location());
-			if (!ssl.infos.empty())
+			if (identifier.name.str() == "blobhash" && !m_evmVersion.hasBlobhash())
+				m_errorReporter.warning(
+					7527_error,
+					nativeLocationOf(identifier),
+					"\"blobhash\" was introduced as builtin function in EVM version Cancun "
+					"but you are currently using EVM version " + m_evmVersion.name() +
+					" and it will not behave as expected for EVM version >= Cancun."
+				);
```

Comment:
This now warns on a *reference* to the name ``blobhash`` for < Cancun, right? That's error cases for evm version < cancun unless there is a user-declaration of ``blobhash``. Shouldn't we rather just let that error happen and have a shadowing warning on such declarations? (I haven't checked how we did this for ``basefee`` in the past)

Reply:
Ah right. Where should the shadowing warning goes then? I didn't see such test in the basefee PR: https://github.com/ethereum/solidity/pull/11763

Reply:
We probably don't need a manual shadowing warning at all - from a quick glance we also didn't do it before and don't do it for the other new opcodes, so it's probably just fine to remove this (you can check the PRs for the other opcodes, we can do it the same way for all of them)

Reply:
But as blobhash is now global and differs from other opcodes, i.e. it is not a member of the magic type like `block.blobbasefee`, for example. It seems we'll need a way to allow its declaration for EVM `<cancun` without causing breaking changes, right? This was the idea behind the condition in the code above.

Consider this example: https://github.com/ethereum/solidity/blob/4c3e11850419fc945e90780a86ab973bc634f912/test/libsolidity/semanticTests/inlineAssembly/blobhash_pre_cancun.sol. This test works well in part 1 of the blobhash PR. However, here, if I remove the condition above, an error is thrown with the message: `This declaration shadows a declaration outside the inline assembly block.` The issue arises from this line: https://github.com/ethereum/solidity/blob/4c3e11850419fc945e90780a86ab973bc634f912/libsolidity/analysis/ReferencesResolver.cpp#L415

This doesn't happen with `blobbasefee` because it is not directly exposed at the solidity level; it is only accessible as a member of `block`. So I'm not sure how to deal with this case. Should this not be allowed for blobhash anyway?

Reply:
I think that the problem here is that we enter the `if` at all on EVM versions `<cancun`.

On `cancun` the error is what we want - you should not be able to shadow it at all, just like you can't shadow `blockhash`. On earlier versions `blobhash` should not be included in the names `nameFromCurrentScope()` returns so there should be no error or warning (`blobash` does not exist on those versions so there's nothing to shadow after all.

Reply:
Got it. Thanks, I will update the code to behave like you described ;)